### PR TITLE
Support SQL Server binary data types

### LINKED ISSUE #386: Add missing SQL Server data types
Binary types: 
- ROWVERSION
- TIMESTAMP
- BINARY
- VARBINARY
- IMAGE

### PR DESCRIPTION
**Pull request by MMatten**
Provide base class for binary data type support.

Provide support for the following SQL Server data types: -
* BINARY
* VARBINARY
* TIMESTAMP
* ROWVERSION

Resolves #386.

### CODE REVIEW

**Comment by javornikolov** [comment]
@MMatten, builds should work now. I am not sure whether it might need manual triggering for an existing pr or branch.

**Comment by MMatten** [comment]
Thanks @javornikolov I've had a (brief) search around and I'm still not quite sure if manually triggering builds for existing PRs/pushes is supported. I've got some tweaks to add to this PR anyway so I'll see what happens then.

**Comment by MMatten** [comment]
I'll ensure that we close #386 as part of this too.

**Comment by javornikolov** [comment]
> Thanks @javornikolov I've had a (brief) search around and I'm still not quite sure if manually triggering builds for existing PRs/pushes is supported. I've got some tweaks to add to this PR anyway so I'll see what happens then.

Maybe pushing something to the branch or force-pusshing the same...

**Comment by MMatten** [comment]
@javornikolov thanks. I've pushed more changes and we're building again. Will be interesting to see how fast we're going to be burning build credits. I'll check shortly.

**Comment by MMatten** [comment]
@javornikolov I'm seeing some new JUnit deprecation warnings. I'll look to deal with these on a different PR (maybe bump up to JUnit 5 too on that one).

I have to at least test this change across all of the adapters as there's that change to the core before marking this as ready for a review.

**Comment by MMatten** [comment]
@javornikolov I've now tested this across all of the adapters, apart from Snowflake as I have no usable environment (I have reached out to Snowflake to see if they'll offer us any long-term support though).

**Comment by MMatten** [comment]
@javornikolov one observation on the Snowflake situation...it seems that we didn't inherit any tests for Snowflake anyway.

**Comment by MMatten** [comment]
@javornikolov any review comments or do you think we can merge this (after any rebasing)?

File: dbfit-java/core/src/main/java/dbfit/util/Binary.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        return binaryData;
+    }
+
+    public static Object parse(String s) throws Exception {
+        List<Byte> bytes = new ArrayList<Byte>();
+        if (!s.substring(0, 2).equals("0x")) {
+            throw new NumberFormatException("Hexadecimal string does not start with 0x");
+        }
+        String input = s.substring(2);
+        for (int i = ((input.length()) / 2); i > 0; i--) {
```
Comment (javornikolov) [inline dbfit-java/core/src/main/java/dbfit/util/Binary.java:31]: I think Apache Commons Codec library already includes hex parsing functionality. Maybe we can use it here?

**Comment by javornikolov** [comment]
> @javornikolov any review comments or do you think we can merge this (after any rebasing)?

@MMatten - looks reasonably good to me. Just see if we can reuse hex parsing, etc. from a library which we already use - as per my comment.

You can push to the PR to trigger a build (sqlserver tests are already available in our CI).

File: dbfit-java/core/src/main/java/dbfit/util/Binary.java (hunk 1 above)
Comment (javornikolov) [reply dbfit-java/core/src/main/java/dbfit/util/Binary.java:31]: It's good to add some unit tests for this.

File: dbfit-java/core/src/main/java/dbfit/util/Binary.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public String toString() {
+        String out = "0x";
+        for (int i = binaryData.length - 1; i >= 0; i--) {
+            out = out + String.format("%02X", binaryData[i]);
+        }
+        return out;
+    }
+
+    @Override
+    public boolean equals(final Object other) {
```
Comment (javornikolov) [inline dbfit-java/core/src/main/java/dbfit/util/Binary.java:53]: That's similar to `Arrays.equals(byte[], byte[])` I think so I got wondering - maybe we can delegate to it?

Comment (MMatten) [reply dbfit-java/core/src/main/java/dbfit/util/Binary.java:37]: Yes. Good shout. Update applied.

**Comment by MMatten** [comment]
@javornikolov so apart from squashing commits, etc, any other things you've spotted? We could always do with more unit tests of course!

File: dbfit-java/core/src/main/java/dbfit/util/Binary.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        if (other == null)
+            return false;
+        if (!(other instanceof Binary)) {
+            return false;
+        }
+        if (!Arrays.equals(binaryData, ((Binary) other).getBytes())) {
```
Comment (javornikolov) [inline dbfit-java/core/src/main/java/dbfit/util/Binary.java:43]: We can directly `return Arrays.equals(...)`. The extra if just adds more complexity and is not needed.

**Comment by MMatten** [comment]
@javornikolov updates applied. I think we could be ready to merge following a final review (please).

**Comment by javornikolov** [comment]
> @javornikolov updates applied. I think we could be ready to merge following a final review (please).

@MMatten, yes, it looks good to me to merge it. 👍

**Pull request merged**